### DISCUSSION TITLE
Duplicated data will cause exhausted computation in milvus?

### DISCUSSION THREAD

**Discussion by xiaojuan8** [milvus-io/milvus#15689] [Q&A and General discussion]
Hi，I found that duplicated data leads to milvus indexing failed.
my data is organized as follows:
1. 22411093 * 32dim vectors
2. insert them into milvus-server repeatedly for 20 loops.
3. create an IVFPQ index (m=8, nlist=2*sqrt(22411093));
4. search 5 items from this index (nprobe=10)
search cost totally 51s. But 22411093 data cost only 10ms on searching 5 items (nprobe=10).

Also, based on this case, I tried (nprobe=100, 1000, 10000, 65535) many times, the total cost on search 5 items is about 50s all the way. I guess that nlist of clusters is not uniformly which caused this situation. So, what is the strategy of indexing for duplicated data, or it is a bug of milvus-server?

----------------------------------
chinese version:
中文版(防止英文表述疏漏)
您好，我发现重复数据在milvus中聚类会有一些问题。
我将我的数据做了以下处理(重复插入的原因是，我们希望测试亿级别的数据性能，但暂时无法得到真实的亿级大规模数据，所以重复数据代替)
1. 22411093 * 32维向量
2. 重复循环20次将以上数据插入milvus-server
3. 创建IVFPQ索引（参数m=8,nlist=2*sqrt(22411093)）
4. 查找5条相似的向量(nprobe=10)
查找总耗时51秒，但是原始22411093的数据才10毫秒不到，即使线性膨胀，也不至于几千倍的耗时膨胀，非常奇怪。
基于以上现象，我们做了以下尝试，将nprobe查询参数遍历(10~65530,做了一些测试)，发现更改nprobe参数，查询总耗时不变，因此怀疑nlist个聚类数据分布非常不均匀，导致计算查找总需要几乎全量数据的计算(接近于暴力计算)，这是非常不符合预期的，所以想咨询一下对于这样重复数据，查询milvus-server是有什么机制？ 还是说这个问题本身就是一个bug？

**Comment by yhmo** [comment]
Some possible reasons:
- If it is the first-time search, milvus need to load data from disk files, the time may waste in I/O. You can call load_collection() before search() to make sure the data is loaded into memory.
- The cache size is not enough to hold the collection in memory. To get the best search performance, we suggest user to set a reasonable cache size in the server_config.yaml:
```
cache:
  cache_size: 4GB
```
22411093 * 32 * 20 = 14343099520 bytes = 13.3 GB, set cache_size = 14GB if without index.
If all the index are finished, the index size should be 13.3 * 0.3 = 4 GB. Set cache_size = 5GB.
If the cache_size is less than the required data size, milvus will swap data between disk and memory, you will get poor performance.

- Index work is not finished. Milvus split data into segments, and build indexes for segments one by one. If index work is not finished and you call search(), milvus will do brute-force search for the un-indexed segments. Use get_collection_stat() to observe segment state:
```
status, info = MILVUS.get_collection_stats(collection_name=name)
print(info)
```
For example, a collection with one segment will return this info in JSON format:
```
{
	'partitions': [{
		'row_count': 10000,
		'segments': [{
			'data_size': 5200000,
			'index_name': 'IDMAP',
			'name': '1645512680644228000',
			'row_count': 10000
		}],
		'tag': '_default'
	}],
	'row_count': 10000
}
```

If the "index_name" is "IDMAP", that means this segment is not indexed.

**Reply by xiaojuan8** [reply]
Hi, thanks, this might be working. I am trying to set cachesize to 14GB.
but OOM occurred while creating an index: https://github.com/milvus-io/milvus/discussions/15697

**Reply by xiaojuan8** [reply]
Hi, reseting cachesize works for me, search latency is reduced to xx ms.
